src/devices/PiJuice/PiJuice.cs

```diff
@@ -0,0 +1,252 @@
+            if (delay > 0)
+            {
+                DelayHelper.DelayMilliseconds(delay, false);
+            }
+
+            var response = ReadCommand(command, (byte)data.Length);
+
+            if (!data.SequenceEqual(response))
+            {
+                throw new IOException($"{nameof(WriteCommandVerify)}: Failed to write command {command}");
+            }
+        }
+
+        /// <summary>
+        /// Read data from PiJuice
+        /// </summary>
+        /// <param name="command">The PiJuice command</param>
+        /// <param name="length">The length of the data to be read</param>
+        /// <returns>Returns an array of the bytes read</returns>
+        internal byte[] ReadCommand(PiJuiceCommand command, byte length)
+        {
+            byte tries = 0;
+            Span<byte> outArray = stackalloc byte[length + 1];
+            outArray.Clear();
+
+            // When writing/reading the I2C port, PiJuice doesn't respond on time in some cases
+            // So we wait a little bit before retrying
+            // In most cases, the I2C read/write can go thru without waiting
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    _i2cDevice.Write(new byte[] { (byte)command });
+                    _i2cDevice.Read(outArray);
+
+                    var checksum = GetCheckSum(outArray);
+                    if (checksum != outArray[length])
+                    {
+                        outArray[0] |= 0x80;
+                        checksum = GetCheckSum(outArray);
+                        if (checksum != outArray[length])
+                        {
+                            tries++;
+
+                            if (tries >= MaxRetries)
+                            {
+                                throw new InvalidDataException($"{nameof(ReadCommand)}: Invalid checksum read command {command}, checksum {checksum}");
+                            }
+
+                            DelayHelper.DelayMilliseconds(ReadRetryDelay, false);
+                            continue;
+                        }
+                    }
+
+                    break;
+                }
+                catch (IOException ex)
+                {
+                    tries++;
+
+                    if (tries >= MaxRetries - 1)
+                    {
+                        throw new IOException($"{nameof(ReadCommand)}: Failed to read command {command}", ex);
+                    }
+
+                    DelayHelper.DelayMilliseconds(ReadRetryDelay, false);
+                }
+            }
+
+            return outArray.Slice(0, length).ToArray();
+        }
+
+        /// <summary>Gets the check sum.</summary>
+        /// <param name="data">The data.</param>
+        /// <param name="all">Whether the last byte in the data is included in the checksum</param>
+        /// <returns>Checksum</returns>
+        private byte GetCheckSum(ReadOnlySpan<byte> data, bool all = false)
```

Comment:
you should rename `all` to `checkLastByte`

Reply:
Will do